### DISCUSSION TITLE
CloudStack Dev Setup.

### DISCUSSION THREAD

**Discussion by lohitslohit** [apache/cloudstack#9157] [Q&A] [answered]
Hi, I am new to CloudStack, How to get started in it. https://docs.cloudstack.apache.org/en/4.19.0.0/developersguide/developer_guide.html

Whether above Docs is okay to get start on it. In above docs unable to download the devCloud.
![image](https://github.com/apache/cloudstack/assets/59111894/7c1d962a-3bd3-4b1f-a994-b1f9ff1bf737)

**Comment by sureshanaparti** [comment, marked as answer]
Hi @lohitslohit Can you try with mbx-based env, details here: https://github.com/shapeblue/hackerbook/blob/main/2-dev.md#mbx-based-development

**Comment by lohitslohit** [comment]
okay. But during the dev environment setup. I had executed the 

`mvn -Pdeveloper,systemvm clean install`

Given in this docs https://docs.cloudstack.apache.org/en/4.19.0.1/developersguide/developer_guide.html
But I had faced issue given below

```
[INFO] Building Apache CloudStack Plugin - ACL Project Role Based 4.20.0.0-SNAPSHOT [47/143]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ cloud-plugin-acl-project-role-based ---
[INFO] Deleting /home/lohit18/OpenSource/cloudstack/Org/cloudstack/plugins/acl/project-role-based (includes = [target, dist], excludes = [])
[INFO]
[INFO] --- maven-checkstyle-plugin:3.1.0:check (cloudstack-checkstyle) @ cloud-plugin-acl-project-role-based ---
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fa68994f0ad, pid=20757, tid=20783
#
# JRE version: OpenJDK Runtime Environment (11.0.19+7) (build 11.0.19+7-post-Ubuntu-0ubuntu118.04.1)
# Java VM: OpenJDK 64-Bit Server VM (11.0.19+7-post-Ubuntu-0ubuntu118.04.1, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x7760ad]  void objArrayOopDesc::oop_iterate_range<G1ScanEvacuatedObjClosure>(G1ScanEvacuatedObjClosure*, int, int)+0xcd
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /home/lohit18/OpenSource/cloudstack/Org/cloudstack/core.20757)
#
# An error report file with more information is saved as:
# /home/lohit18/OpenSource/cloudstack/Org/cloudstack/hs_err_pid20757.log
#
# If you would like to submit a bug report, please visit:
#   https://bugs.launchpad.net/ubuntu/+source/openjdk-lts
#
Aborted (core dumped)
```

How to fix this issue.

**Comment by weizhouapache** [comment]
> okay. But during the dev environment setup. I had executed the 
> 
> `mvn -Pdeveloper,systemvm clean install`
> 
> Given in this docs https://docs.cloudstack.apache.org/en/4.19.0.1/developersguide/developer_guide.html
> But I had faced issue given below
> 
> ```
> [INFO] Building Apache CloudStack Plugin - ACL Project Role Based 4.20.0.0-SNAPSHOT [47/143]
> [INFO] --------------------------------[ jar ]---------------------------------
> [INFO]
> [INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ cloud-plugin-acl-project-role-based ---
> [INFO] Deleting /home/lohit18/OpenSource/cloudstack/Org/cloudstack/plugins/acl/project-role-based (includes = [target, dist], excludes = [])
> [INFO]
> [INFO] --- maven-checkstyle-plugin:3.1.0:check (cloudstack-checkstyle) @ cloud-plugin-acl-project-role-based ---
> #
> # A fatal error has been detected by the Java Runtime Environment:
> #
> #  SIGSEGV (0xb) at pc=0x00007fa68994f0ad, pid=20757, tid=20783
> #
> # JRE version: OpenJDK Runtime Environment (11.0.19+7) (build 11.0.19+7-post-Ubuntu-0ubuntu118.04.1)
> # Java VM: OpenJDK 64-Bit Server VM (11.0.19+7-post-Ubuntu-0ubuntu118.04.1, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)
> # Problematic frame:
> # V  [libjvm.so+0x7760ad]  void objArrayOopDesc::oop_iterate_range<G1ScanEvacuatedObjClosure>(G1ScanEvacuatedObjClosure*, int, int)+0xcd
> #
> # Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /home/lohit18/OpenSource/cloudstack/Org/cloudstack/core.20757)
> #
> # An error report file with more information is saved as:
> # /home/lohit18/OpenSource/cloudstack/Org/cloudstack/hs_err_pid20757.log
> #
> # If you would like to submit a bug report, please visit:
> #   https://bugs.launchpad.net/ubuntu/+source/openjdk-lts
> #
> Aborted (core dumped)
> ```
> 
> How to fix this issue.

jvm heap size?

pls: ubuntu 18.04 is eol.

**Comment by lohitslohit** [comment]
@weizhouapache  Again I had tried to run the same cmd. but I got amother error. 

```
[INFO] Apache CloudStack Developer Tools .................. SUCCESS [  0.014 s]
[INFO] Apache CloudStack apidocs .......................... FAILURE [  7.339 s]
[INFO] Apache CloudStack marvin ........................... SKIPPED
[INFO] Apache CloudStack DevCloud4 ........................ SKIPPED
[INFO] Apache CloudStack DevCloud-KVM ..................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  54:46 min
[INFO] Finished at: 2024-05-25T19:59:23+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (compile) on project cloud-apidoc: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :cloud-apidoc
lohit18@lohit18-Standard-PC-Q35-ICH9-2009:~/OpenSource/cloudstack/Org/cloudstack$
```

How to get the jvm size ? 
I am using the ubuntu 18.04. Is it okay to use 18.04

**Comment by weizhouapache** [comment]
> @weizhouapache  Again I had tried to run the same cmd. but I got amother error. 
> 
> ```
> [INFO] Apache CloudStack Developer Tools .................. SUCCESS [  0.014 s]
> [INFO] Apache CloudStack apidocs .......................... FAILURE [  7.339 s]
> [INFO] Apache CloudStack marvin ........................... SKIPPED
> [INFO] Apache CloudStack DevCloud4 ........................ SKIPPED
> [INFO] Apache CloudStack DevCloud-KVM ..................... SKIPPED
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD FAILURE
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time:  54:46 min
> [INFO] Finished at: 2024-05-25T19:59:23+05:30
> [INFO] ------------------------------------------------------------------------
> [ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (compile) on project cloud-apidoc: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
> [ERROR]
> [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
> [ERROR] Re-run Maven using the -X switch to enable full debug logging.
> [ERROR]
> [ERROR] For more information about the errors and possible solutions, please read the following articles:
> [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
> [ERROR]
> [ERROR] After correcting the problems, you can resume the build with the command
> [ERROR]   mvn <goals> -rf :cloud-apidoc
> lohit18@lohit18-Standard-PC-Q35-ICH9-2009:~/OpenSource/cloudstack/Org/cloudstack$
> ```
> 
> How to get the jvm size ? 
> I am using the ubuntu 18.04. Is it okay to use 18.04 

As I remember, this is caused by a missing program (I forget its name), you can check the output of the compilation

**Comment by lohitslohit** [comment]
okay @weizhouapache 

How to check the JVM heap size?
I am using the ubuntu 18.04. Is it okay to use 18.04

**Comment by weizhouapache** [comment]
> okay @weizhouapache
> 
> How to check the JVM heap size? I am using the ubuntu 18.04. Is it okay to use 18.04

export MAVEN_OPTS="-Xmx4096m -XX:MaxMetaspaceSize=1024m"
then run the maven command
it may be not needed, as your build already passed. 
the error with `apidocs` project is not caused by java settings or any java codes.

**Comment by yadvr** [comment]
@lohitslohit you're better off following this https://github.com/shapeblue/hackerbook/blob/main/2-dev.md
The upstream docs website is outdated and we need to improve it. Ubuntu 18.04 isn't support yet, it might still work OK as a dev-testing host though, I would recommend that you use/try 22.04 or 20.04.
